### ISSUE TITLE
Should nested classes be allowed to be defined in universal traits?

### ISSUE THREAD

**Issue by odersky** [lampepfl/dotty#18237]
## Compiler version

3.3.2

## Minimized example

```scala
trait A extends Any:
  class C
```

## Output

```scala
-- Error: ../../new/test.scala:2:8 ---------------------------------------------
2 |  class C
  |  ^^^^^^^
  |  class C cannot be defined in universal trait A
```

## Expectation

Not sure. Scala 2 allows this, and the Scala 2 standard library has several places where this occurs. 

But the problem is that value classes cannot define nested classes. Indeed the [value class spec](https://docs.scala-lang.org/overviews/core/value-classes.html) says

> A value class …
> … can only have defs as members. In particular, it cannot have lazy vals, vars, or vals as members.

Since value classes can extend universal traits it seems that universal traits should also be forbidden to have value classes.
So should we close this issue here and file an issue against Scala 2 instead? /cc @lrytz, @dwijnand @SethTisue.

**Comment by odersky** [comment]
I did some more digging into the stdlib issues and found out that one cannot simply move nested classes in the universal trait `SeqOps` out of the trait, since they access several protected methods in `SeqOps`. Which seems to indicate that
at runtime we need an outer pointer from these classes to the `SeqOps` instance. But if `SeqOps` was indeed implemented with a value class, such an outer pointer could not be generated.

Which begs the question: Has anyone tried to implement `SeqOps` with a value class? (I would suppose not). Therefore, it seems that `SeqOps` and all its subclasses really should extend `AnyRef` instead of `Any`. Would this be a minary compatible change? /cc @sjrd @nicolasstucki

**Comment by odersky** [comment]
Also /cc @julienrf @szeiger

**Comment by julienrf** [comment]
I don’t think that would be a binary incompatible change (I know Sébastien will correct me if I am wrong), but that will be a source incompatible change. I guess the only reason to have them extend `Any` instead of `AnyRef` is to allow collection implementations to be value classes. None of the standard collections are value classes, but any existing implementation of collection in the wild that is a value class will not compile anymore.

I experimented to make `SeqOps` extend `AnyRef` instead of `Any` in https://github.com/scala/scala/pull/10470

**Comment by odersky** [comment]
But it seems that the classes nested in SeqOps would not work correctly for value class implementations of Seq because they would require an outer pointer, and value classes cannot be referred by outer pointers. So has anyone tried that? Implement Seq with a valueclass and then do an operation that requires a `PermutationItr` or a `CombinationItr`? I'd be surprised if that worked.

**Comment by julienrf** [comment]
Ah good point. I just tried and it seems to be working: https://github.com/scala/scala/pull/10471

The test creates an instance of the collection and calls `.permutations` on it. Finally, it prints the following:

> List(ArrayWrapper(1, 2, 3), ArrayWrapper(1, 3, 2), ArrayWrapper(2, 1, 3), ArrayWrapper(2, 3, 1), ArrayWrapper(3, 1, 2), ArrayWrapper(3, 2, 1))

**Comment by nicolasstucki** [comment]
> Has anyone tried to implement SeqOps with a value class?

I have not tried that.

**Comment by dwijnand** [comment]
I say we just allow universal traits to define nested classes.  And if need be, make defining of a value class that extends such a trait not compile.

**Comment by lrytz** [comment]
> SeqOps would not work correctly for value class implementations of Seq because they would require an outer pointer, and value classes cannot be referred by outer pointers. So has anyone tried that? Implement Seq with a valueclass and then do an operation that requires a `PermutationItr` or a `CombinationItr`? I'd be surprised if that worked.

In Scala 2 at least, calling an operation on a universal trait always boxes (https://github.com/scala/bug/issues/9731). So

```scala
trait T extends Any {
  def a = "hi"
  class C { def b = a }
  def c = new C().b
}
```

works because `T.a` always requires an instance of `T`.

@sjrd points out that changing `Ops` traits to be non-universal is not source compatible, it would break existing subclasses.

**Comment by odersky** [comment]
> In Scala 2 at least, calling an operation on a universal trait always boxes (https://github.com/scala/bug/issues/9731). 

Yes, and that's an open bug, right?

> @sjrd points out that changing Ops traits to be non-universal is not source compatible, it would break existing subclasses.

Which subclasses would break? Only value classes or also others?

**Comment by odersky** [comment]
OK, so what happens in Scala 2 if a universal trait contains nested classes and is extended by a value class?

**Comment by lrytz** [comment]
The class takes an `$outer` instance:

```scala
trait U extends Any {
  def a = 1
  class C { def b = a }
  def c = (new C).b // new U$C(this).b
}

class K(val m: String) extends AnyVal with U {
  def d = (new C).b // object K { def d$extension($this: String): Int = new U$C(new K($this)).b() }
}
```

**Comment by odersky** [comment]
Ah, I think I understand better now. We should try to lift the restriction in Scala 3 and find out whether the behavior would be the same.
